filters/include/pcl/filters/lambda_filter_indices.h

```diff
@@ -0,0 +1,166 @@
+                                           Function,
+                                           const pcl::remove_cvref_t<PointT>&,
+                                           pcl::index_t>,
+                     bool>;
+
+// can't use this for SFINAE since Derived isn't properly defined
+// but this can be used after the class definition to test it
+/*
+template <class Base, class Derived>
+constexpr auto IsValidLambdaFilter = std::enable_if_t<
+    std::is_base_of<Base, Derived>::value &&
+        pcl::is_invocable_v<std::declval<Derived>().get_lambda, void> &&
+        std::is_same<PointFilterLambda<std::declval<Derived>().get_lambda()>,
+bool>, bool>;
+*/
+
+/**
+ * \brief LambdaFilterIndices filters point clouds and indices based on a
+ * function pointer passed to filter command \ingroup filters
+ */
+template <typename PointT, typename Derived>
+struct LambdaFilterIndicesImpl : public FilterIndices<PointT> {
+private:
+  using Base = FilterIndices<PointT>;
+  using PCLBase = pcl::PCLBase<PointT>;
+
+protected:
+  using PointCloud = typename FilterIndices<PointT>::PointCloud;
+  using PointCloudPtr = typename PointCloud::Ptr;
+  using PointCloudConstPtr = typename PointCloud::ConstPtr;
+  using FieldList = typename pcl::traits::fieldList<PointT>::type;
+
+  using Base::extract_removed_indices_;
+  using Base::removed_indices_;
+  using PCLBase::indices_;
+  using PCLBase::input_;
+
+protected:
+  // to prevent instantiation of impl class
+  LambdaFilterIndicesImpl(bool extract_removed_indices)
+  : FilterIndices<PointT>(extract_removed_indices)
+  {}
+  /** \brief Filtered results are indexed by an indices array.
+   * \param[out] indices The resultant indices.
+   */
+  void
+  applyFilter(std::vector<int>& indices) override
+  {
+    indices.clear();
+    indices.reserve(input_->points.size());
+    if (extract_removed_indices_) {
+      removed_indices_->clear();
+      removed_indices_->reserve(input_->points.size());
+    }
+
+    const auto& lambda = static_cast<Derived*>(this)->get_lambda();
+
+    for (const auto index : *indices_) {
+      // lambda returns true for points that shoud be selected
+      if (!(this->negative_) == lambda(input_->points[index], index)) {
+        indices.push_back(index);
+      }
+      else {
+        if (extract_removed_indices_) {
+          removed_indices_->push_back(index);
+        }
+      }
+    }
+  }
+};
+} // namespace detail
+
+template <typename PointT, typename Functor>
+struct LambdaFilterIndices
+: public detail::LambdaFilterIndicesImpl<PointT, LambdaFilterIndices<PointT, Functor>> {
+private:
+  using Self = LambdaFilterIndices<PointT, Functor>;
+  using Base = detail::LambdaFilterIndicesImpl<PointT, Self>;
+
+public:
+  using FunctorT = Functor;
+  // using in type would complicate signature
+  static_assert(std::is_same<detail::PointFilterLambda<PointT, FunctorT>, bool>::value,
+                "Functor needs to be able to satisfy the callable constraint");
```

Comment:
This might be cryptic to interpret for C++ beginners. Something along the lines of "The callable object needs to have the following signature bool(const PointT&, index_t)" gives more information.